### ISSUE TITLE
Support for Next.js 15 RC and React 19 RC

### ISSUE THREAD

**Issue by Aralf** [amannn/next-intl#1075]
### Description

I have taken the example code from `example-app-router-next-auth` that has the With i18n Routing setup; and tried to use latest Next canary (18.3.0-canary.70 at the time of writing this) with React 19 RC in order to test both that and the new React compiler.

There seems to be an issue with React 19, as the application fails to run, even without trying the React compiler.

Without turbo it shows the following error.
![error without turbo](https://github.com/amannn/next-intl/assets/7780231/17216932-b918-4e5b-8e17-84fbf94b60c0)

With turbo, instead shows this error.
![error with turbo](https://github.com/amannn/next-intl/assets/7780231/dfed643c-0e62-4a5c-81f5-3fbb12f27101)

I have made another branch where I have follow the documentation for `Without i18n routing`. This removes the `[locale]`, `navigation.ts` and use of next-intl middleware on `middleware.ts`. With these changes the application runs perfectly fine both with/without turbo and the react compiler.

### Verifications

- [X] I've verified that the problem I'm experiencing isn't covered in [the docs](https://next-intl-docs.vercel.app/).
- [X] I've searched for similar, existing issues on GitHub and [Stack Overflow](https://stackoverflow.com/search?q=next-intl).
- [X] I've compared my app to [a working example](https://next-intl-docs.vercel.app/examples) to look for differences.

### Mandatory reproduction URL

https://github.com/Aralf/next-intl-react-19-compiler

### Reproduction description

Steps to reproduce:
1. Run `next dev` with or without turbo
2. It will not work, in both cases throws different errors.
3. There is no need for trying React compiler as it already fails.
4. Change to branch `without-i18n-routing` 
5. React compiler is already enabled on next.config.mjs
6. Run `next dev` with or without turbo
7. It works woks for all cases

### Expected behaviour

It should work for both the With i18n routing and the Without i18n routing setups.

**Comment by amannn** [comment]
Hmm, that's interesting. The Next.js 14 release is scheduled for next week as far as I know, maybe worth waiting until then. It might be that they're still ironing out some details.

**Comment by heresyrj** [comment]
can confirm this issue. 
reproduced the same error on my project.

**Comment by Gawdfrey** [comment]
Able to reproduce it here as well.
As mentioned by @amannn, probably worth waiting until [next 15 is launched this week](https://x.com/feedthejim/status/1791246068971381246) until further investigation.

**Comment by singuerinc** [comment]
Hi, I just tried to update to Next.js 15 RC and it breaks with:

```
 ⨯ ../src/shared/lib/router-context.shared-runtime.ts (4:8) @ <unknown>
 ⨯ react__WEBPACK_IMPORTED_MODULE_0__.createContext is not a function
  2 | import type { NextRouter } from './router/router'
  3 |
> 4 | export const RouterContext = React.createContext<NextRouter | null>(null)
    |        ^
  5 |
  6 | if (process.env.NODE_ENV !== 'production') {
  7 |   RouterContext.displayName = 'RouterContext'
 ⚠ ./node_modules/next-intl/dist/esm/navigation/shared/BaseLink.js
```

**Comment by zipme** [comment]
This is the error I got:
```
./node_modules/.pnpm/next-intl@3.11.3_next@15.0.0-rc.0_react@19.0.0-rc-4c2e457c7c-20240522/node_modules/next-intl/dist/esm/navigation/shared/BaseLink.js
Attempted import error: 'useState' is not exported from 'react' (imported as 'i').

Import trace for requested module:
./node_modules/.pnpm/next-intl@3.11.3_next@15.0.0-rc.0_react@19.0.0-rc-4c2e457c7c-20240522/node_modules/next-intl/dist/esm/navigation/shared/BaseLink.js
./node_modules/.pnpm/next-intl@3.11.3_next@15.0.0-rc.0_react@19.0.0-rc-4c2e457c7c-20240522/node_modules/next-intl/dist/esm/navigation/react-server/ServerLink.js
./node_modules/.pnpm/next-intl@3.11.3_next@15.0.0-rc.0_react@19.0.0-rc-4c2e457c7c-20240522/node_modules/next-intl/dist/esm/navigation/react-server/createSharedPathnamesNavigation.js
./node_modules/.pnpm/next-intl@3.11.3_next@15.0.0-rc.0_react@19.0.0-rc-4c2e457c7c-20240522/node_modules/next-intl/dist/esm/navigation.react-server.js
````

**Comment by amannn** [comment]
I've just upgraded two examples in https://github.com/amannn/next-intl/pull/1089 to Next.js 15 RC and React 19 RC. Generally this seems to work, but it looks like the issue you are seeing comes from https://github.com/vercel/next.js/pull/65424.

Currently in the docs we suggest to import i18n config from `./navigation` into the middleware, where also navigation APIs are defined. These navigation APIs import from `next/link` in turn, which results in the errors you're seeing.

For the time being you can fix this by making sure you're not importing anything from the navigation module into the middleware (see https://github.com/amannn/next-intl/pull/1089/commits/b7469596d953f113e6f63a57f749db341f817f68).

I've [asked the Next.js team for clarification](https://github.com/vercel/next.js/pull/65424#issuecomment-2128902585) if we should adjust the docs here.

**Comment by amannn** [comment]
Side note for those living on the bleeding edge: The test suite from `next-intl` discovered that serving sitemaps might be broken in Next.js 15 RC (see https://github.com/vercel/next.js/pull/65507#issuecomment-2128974838)

**Comment by Gawdfrey** [comment]
> I've just upgraded two examples in #1089 to Next.js 15 RC and React 19 RC. Generally this seems to work, but it looks like the issue you are seeing comes from [vercel/next.js#65424](https://github.com/vercel/next.js/pull/65424).
> 
> Currently in the docs we suggest to import i18n config from `./navigation` into the middleware, where also navigation APIs are defined. These navigation APIs import from `next/link` in turn, which results in the errors you're seeing.
> 
> For the time being you can fix this by making sure you're not importing anything from the navigation module into the middleware (see [b746959](https://github.com/amannn/next-intl/commit/b7469596d953f113e6f63a57f749db341f817f68)).
> 
> I've [asked the Next.js team for clarification](https://github.com/vercel/next.js/pull/65424#issuecomment-2128902585) if we should adjust the docs here.

Cheers, moved them to a separate file for now, works excellently!

**Comment by Aralf** [comment]
> I've just upgraded two examples in #1089 to Next.js 15 RC and React 19 RC. Generally this seems to work, but it looks like the issue you are seeing comes from [vercel/next.js#65424](https://github.com/vercel/next.js/pull/65424).
> 
> Currently in the docs we suggest to import i18n config from `./navigation` into the middleware, where also navigation APIs are defined. These navigation APIs import from `next/link` in turn, which results in the errors you're seeing.
> 
> For the time being you can fix this by making sure you're not importing anything from the navigation module into the middleware (see [b746959](https://github.com/amannn/next-intl/commit/b7469596d953f113e6f63a57f749db341f817f68)).
> 
> I've [asked the Next.js team for clarification](https://github.com/vercel/next.js/pull/65424#issuecomment-2128902585) if we should adjust the docs here.

Thanks, that is clearly the issue. I have made the change on my reproduction with i18n navigation and it works.

Funny enough, I checked on my company app and noticed that in there I had already separated in two files the locales configuration and the navigation, and never imported navigation to the middleware. So it shouldn't have failed....unless the issue is somewhere else.

Well, turns out Auth.js v5 has the same issue. My reproduction uses v4 as that is the same from next-intl examples. I tested to upgrade to v5 and voilà, again the same error. I suppose I will have to open an issue on next-auth now 😅

Thanks again @amannn for finding the root of the issue.

**Comment by Aralf** [comment]
I've been pointed out [here](https://github.com/nextauthjs/next-auth/issues/11006#issuecomment-2130793154) latest Next [canary](https://github.com/vercel/next.js/releases/tag/v14.3.0-canary.81) includes this [fix](https://github.com/vercel/next.js/pull/66175) that resolves all these issues caused by https://github.com/vercel/next.js/pull/65424 
I've tested it on the reproduction and sure enough no more issues.

**Comment by theoludwig** [comment]
Indeed latest Next.js canary fixes the middleware issue.

However, with the update to Next.js v15 and React.js v19, another issue arise:
When using [PPR](https://www.youtube.com/watch?v=MTcPrTIBkpA), and using both static and dynamic components, with a page using: `unstable_setRequestLocale(props.params.locale)`, it will throw an error when navigating to the page client side (e.g: a link):

```js
Error: Dynamic server usage: Route / couldn't be rendered statically because it used unstable_noStore().
See more info here: https://nextjs.org/docs/messages/dynamic-server-error
```

While with Next.js v14, it worked without issues. I'm not sure if this is an issue of Next.js or next-intl.

Example code to reproduce the issue:

Page
```tsx
const HomePage: React.FC<HomePageProps> = (props) => {
  const { params } = props

  // Enable static rendering
  unstable_setRequestLocale(params.locale)

  const t = useTranslations()
  
  return (
    <main>
      <h1>{t("home.title")}</h1>
      
      <Suspense fallback={<PostLoadingUI />}>
        <PostRandomData />
      </Suspense>
    </main>
  )
}
```

Component
```tsx
export const PostRandomData: React.FC<PostRandomDataProps> = async () => {
  noStore()
  const randomPostId = getRandomNumber(1, 100)
  const post = await getPostById({ id: randomPostId })
  return <PostUI post={post} />
}
```

**Comment by fernandojbf** [comment]
I'm using `getPathname` on middleware to translate some routes.

<img width="323" alt="image" src="https://github.com/amannn/next-intl/assets/9076868/6686af67-96c2-402c-8605-286c1c1d7c9d">

Since `getPathname` is exported by `next-intl/navigation` would be nice to have something that work only on SSR (middleware and API routes).

**Comment by makoveckib** [comment]
[Braking change in [v15.0.0-canary.171]](https://github.com/vercel/next.js/pull/68812)
change : (https://github.com/vercel/next.js/releases/tag/v15.0.0-canary.171)

**Comment by amannn** [comment]
@makoveckib Yep, I've written a detailed analysis about this in https://github.com/amannn/next-intl/issues/1375 and a canary release is already available with the relevant changes if someone wants to test it: https://github.com/amannn/next-intl/pull/1391.

**Comment by makoveckib** [comment]
> sill getting warnings:  In route /[locale] a header property was accessed directly with `headers().get('X-NEXT-INTL-LOCALE')`. `headers()` should be awaited before using its value. Learn more: https://nextjs.org/docs/messages/sync-dynamic-apis

"next-intl": "^3.21.0-canary.0",
"next": "^15.0.0-canary.176",

**Comment by amannn** [comment]
There’s some minimal migration necessary if you want to get rid of the warnings, please refer to the linked PRs in the release PR.

**Comment by amannn** [comment]
I just gave Next.js 15 RC 2 a shot, and while some things have improved, it seems like there is still a bug on their end: https://github.com/vercel/next.js/issues/69000#issuecomment-2416459189. So be careful if you decide to upgrade to it.

Apart from that, be careful when running the codemod. I noticed the codemod transforms pages to `async` when they use hooks like `useTranslations`—this will break them.

**Broken:**

```tsx
export default async function Page(props) {
  const searchParams = await props.searchParams;
  const t = useTranslations('Page')

  // ...
}
```

**Fixed:**

```tsx
import {use} from 'react';

// (no async)
export default function Page(props) {
  const searchParams = use(props.searchParams);
  const t = useTranslations('Page')

  // ...
}
```

See also the notes in https://github.com/vercel/next.js/pull/68812. The recently released [`eslint-plugin-react-hooks@5`](https://www.npmjs.com/package/eslint-plugin-react-hooks) will also warn for this.

**Comment by wottpal** [comment]
Thanks for your efforts, @amannn? If we want to make the switch, is the latest canary on npm `0.0.0-canary-0ea6000` reflecting that branch or how do we install it?

**Comment by amannn** [comment]
@wottpal Yep, that's the latest one! If you install `next-intl@canary` that should always give you the most recent one.

Regarding `3.22.0-canary.1`: That's a slightly outdated one, I had to adapt the canary publishing process due to some problems with Lerna recently which resulted in a new versioning scheme.

**Comment by wottpal** [comment]
Ty! Got everything running smoothly on v15 🚀

**Comment by jspm2013** [comment]
hi @amannn! Thanks for your great job on all this stuff and a question, about your last mention above regarding "...there's [a critical bug](https://github.com/amannn/next-intl/issues/1442) in Next.js 15...":

Are the docs updated now, or are we still waiting for the solution on this bug?

**Comment by amannn** [comment]
next-intl can surely be updated to the latest version, I only found that Next.js 15 has a bug with alternate links (see above).

Additionally, I found this one the other day: https://github.com/vercel/next.js/issues/71623. That’s not really related to next-intl though, rather a general bug and might not affect your app.

I’ll update the examples and docs in this repo once the bug with alternate links is fixed.
